ledger/participant-integration-api/src/test/suite/scala/platform/store/appendonlydao/events/BufferedTransactionsReaderSpec.scala

```diff
@@ -0,0 +1,205 @@
+    )
+
+    offsetUpdates.foreach { case (offset, update) =>
+      transactionsBuffer.push(offset, update)
+    }
+
+    def readerGetTransactionsGeneric(
+        eventsBuffer: EventsBuffer[Offset, TransactionLogUpdate],
+        startExclusive: Offset,
+        endInclusive: Offset,
+        fetchTransactions: FetchTransactions[Object, String],
+    ) =
+      BufferedTransactionsReader
+        .getTransactions(eventsBuffer)(
+          startExclusive = startExclusive,
+          endInclusive = endInclusive,
+          filter = filter,
+          verbose = false,
+        )(
+          toApiTx = toApiTx,
+          apiResponseCtor = apiResponseCtor,
+          fetchTransactions = fetchTransactions,
+          sourceTimer = metrics.daml.services.index.streamsBuffer.getTransactionTrees,
+          resolvedFromBufferCounter =
+            metrics.daml.services.index.streamsBuffer.transactionTreesBuffered,
+          totalRetrievedCounter = metrics.daml.services.index.streamsBuffer.transactionTreesTotal,
+          bufferSizeCounter = metrics.daml.services.index.transactionTreesBufferSize,
+          outputStreamBufferSize = 128,
+        )
+        .runWith(Sink.seq)
+
+    "request within buffer range inclusive on offset gap as start exclusive" should {
+      "fetch from buffer" in {
+        readerGetTransactionsGeneric(
+          eventsBuffer = transactionsBuffer,
+          startExclusive = predecessor(offset3),
+          endInclusive = offset4,
+          fetchTransactions = (_, _, _, _) => fail("Should not fetch"),
+        ).map(
+          _ should contain theSameElementsInOrderAs Seq(
+            offset3 -> apiResponse3,
+            offset4 -> apiResponse4,
+          )
+        )
+      }
+    }
+
+    "request within buffer range inclusive on existing offset as start inclusive" should {
+      "fetch from buffer" in {
+        readerGetTransactionsGeneric(
+          eventsBuffer = transactionsBuffer,
+          startExclusive = offset2,
+          endInclusive = offset4,
+          fetchTransactions = (_, _, _, _) => fail("Should not fetch"),
+        ).map(
+          _ should contain theSameElementsInOrderAs Seq(
+            offset3 -> apiResponse3,
+            offset4 -> apiResponse4,
+          )
+        )
+      }
+    }
+
+    "request before buffer start" should {
+      "fetch from buffer and storage" in {
+        readerGetTransactionsGeneric(
+          eventsBuffer = transactionsBuffer,
+          startExclusive = offset1,
+          endInclusive = offset3,
+          fetchTransactions = {
+            case (`offset1`, `offset2`, `filter`, false) =>
+              Source.single(offset2 -> apiResponse4)
+            case unexpected => fail(s"Unexpected $unexpected")
+          },
+        ).map(
+          _ should contain theSameElementsInOrderAs Seq(
+            offset2 -> apiResponse4, // apiResponse4 on purpose in order to distinguish from fetched from buffer
```

Comment:
It would be better to construct a special response for this rather than reusing 4

Reply:
Done